coda_v2_python_client/firebase_client_wrapper.py

```diff
@@ -622,3 +622,39 @@ def set_user_ids(self, dataset_id, user_ids):
                 batch.set(self.get_segment_ref(segment_id), {"users": user_ids})
         batch.commit()
         log.debug(f"Wrote {len(user_ids)} users to dataset {dataset_id}")
+
+    def add_message_to_dataset(self, dataset_id, message, MAX_SEGMENT_SIZE=2500):
+        """
+        Adds message to a given dataset.
+
+        :param dataset_id: Id of the dataset to add the message into.
+        :type dataset_id: str
+        :param message: The message to be added.
+        :type message: core_data_modules.data_models.message.Message
+        :param MAX_SEGMENT_SIZE: the maximum size for a segment, defaults to 2500
+        :type MAX_SEGMENT_SIZE: int, optional
+        """
+        existing_segment_messages = dict()  # of segment id -> (dict of message id -> Message)
+        latest_segment_index = self.get_segment_count(dataset_id)
+        latest_segment_id = self.id_for_segment(dataset_id, latest_segment_index)
+        # Note: We need to read the latest segment so we know how big it is, but there's no need to check previous segments.
+        existing_segment_messages[latest_segment_id] = self.get_segment_messages(self.id_for_segment(latest_segment_id))
+
+        latest_segment_size = len(existing_segment_messages[latest_segment_id])
+        if latest_segment_size >= MAX_SEGMENT_SIZE:
+            self.create_next_segment(dataset_id)
+            latest_segment_index = self.get_segment_count(dataset_id)
+            latest_segment_id = self.id_for_segment(dataset_id, latest_segment_index)
+            existing_segment_messages[self.id_for_segment(dataset_id, latest_segment_index)] = []
+
+        batch = self._client.batch()
+
+        message = message.copy()
+        message.last_updated = firestore.firestore.SERVER_TIMESTAMP
+        # Note: Each document costs 2 writes due to the additional write needed by the server to set last_updated
```

Comment:
I think I read somewhere that this was relaxed to a single write, but can't find the release note after a quick search.
In any case, I don't this comment is relevant here as it originally existed to explain why batches appeared to be half-sized.